gcloud/rest/auth/token.py

```diff
@@ -1,95 +1,199 @@
+
+        if self.token_type == Type.GCE_METADATA:
+            self.ensure_token()
+            with self.google_api_lock:
+                resp = self.session.get(GCE_ENDPOINT_PROJECT, timeout=10,
+                                        headers=GCE_METADATA_HEADERS)
+            resp.raise_for_status()
+            project = project or resp.text
+        elif self.token_type == Type.SERVICE_ACCOUNT:
+            project = project or self.service_data.get('project_id')
+
+        return project
+
+    def get(self):
+        # type: () -> str
+        self.ensure_token()
+        return self.access_token
+
+    def ensure_token(self):
+        # type: ()  -> None
+        if not self.access_token:
+            self.acquire_access_token()
+            return
+
+        now = datetime.datetime.utcnow()
+        delta = (now - self.access_token_acquired_at).total_seconds()
+        if delta <= self.access_token_duration / 2:
+            return
 
-        self.age = datetime.datetime.now()
-        self.expiry = 60
-        self.value = None
+        self.acquire_access_token()
 
-    def __str__(self):
-        self.ensure()
-        return str(self.value)
+    def _refresh_authorized_user(self, timeout):
+        # type: (int) -> requests.Response
+        payload = urlencode({
+            'grant_type': 'refresh_token',
+            'client_id': self.service_data['client_id'],
+            'client_secret': self.service_data['client_secret'],
+            'refresh_token': self.service_data['refresh_token'],
+        }, quote_via=quote_plus)
+        with self.google_api_lock:
+            return self.session.post(self.token_uri, data=payload,
+                                     headers=REFRESH_HEADERS, timeout=timeout)
 
-    def assertion(self):
-        with open(self.creds, 'r') as f:
-            credentials = json.loads(f.read())
+    def _refresh_gce_metadata(self, timeout):
+        # type: (int) -> requests.Response
+        with self.google_api_lock:
+            return self.session.get(self.token_uri,
+                                    headers=GCE_METADATA_HEADERS,
+                                    timeout=timeout)
 
-        # N.B. the below exists to avoid using this private method:
-        #   return ServiceAccountCredentials._generate_assertion()
+    def _refresh_service_account(self, timeout):
+        # type: (int) -> requests.Response
         now = int(time.time())
-        payload = {
-            'aud': TOKEN_URI,
-            'exp': now + 3600,
+        assertion_payload = {
+            'aud': self.token_uri,
+            'exp': now + GCLOUD_TOKEN_DURATION,
             'iat': now,
-            'iss': credentials['client_email'],
-            'scope': ' '.join(self.scopes),
+            'iss': self.service_data['client_email'],
+            'scope': self.scopes,
         }
 
-        return jwt.encode(payload, credentials['private_key'],
-                          algorithm='RS256')
-
-    def acquire(self):
```

Comment:
Similarly, lets keep around the old public API aliased to the new version along with `DeprecationWarnings`, something like:
```python
def acquire(self):
    warnings.warn('Token.acquire() is deprecated, please use Token.get()', DeprecationWarning)
```

Reply:
Cool, will do. Should acquire() then also call self.get() in the deprecation period?

Reply:
Basically, we want to match behavior as much as possible. Looks like `acquire()` is closest approximated by `acquire_access_token()`, `ensure()` is roughly equivalent to `ensure_token()`, etc. Basically, we want to make sure we have a few more releases where users aren't _forced_ to change their API so they have time to properly migrate before we do a major version bump with breaking changes.